cores/esp32/esp32-hal-uart.c

```diff
@@ -69,34 +76,45 @@ static uart_t _uart_bus_array[3] = {
 
 static void uart_on_apb_change(void * arg, apb_change_ev_t ev_type, uint32_t old_apb, uint32_t new_apb);
 
-static void IRAM_ATTR _uart_isr(void *arg)
+static uart_interrupt_t *_uart_interrupt_array[3] = {NULL, NULL, NULL};
+
+static void IRAM_ATTR _uart_isr()
 {
     uint8_t i, c;
     BaseType_t xHigherPriorityTaskWoken;
     uart_t* uart;
+    uart_interrupt_t *uart_interrupt;
 
     for(i=0;i<3;i++){
         uart = &_uart_bus_array[i];
-        if(uart->intr_handle == NULL){
+        uart_interrupt = _uart_interrupt_array[i];
+
+        if (uart->intr_handle == NULL) {
             continue;
         }
-        uart->dev->int_clr.rxfifo_full = 1;
-        uart->dev->int_clr.frm_err = 1;
-        uart->dev->int_clr.rxfifo_tout = 1;
-        while(uart->dev->status.rxfifo_cnt || (uart->dev->mem_rx_status.wr_addr != uart->dev->mem_rx_status.rd_addr)) {
+
+        while (uart->dev->status.rxfifo_cnt || (uart->dev->mem_rx_status.wr_addr != uart->dev->mem_rx_status.rd_addr)) {
             c = uart->dev->fifo.rw_byte;
-            if(uart->queue != NULL)  {
+            if (uart_interrupt != NULL && uart_interrupt->dev->num == uart->num && uart_interrupt->func != NULL) {
+                // Fully optimized code would not create the queue anymore if an function has been specified as an argument.
+                (*uart_interrupt->func)(c, uart_interrupt->user_arg);
+            }else if (uart->queue != NULL) {
                 xQueueSendFromISR(uart->queue, &c, &xHigherPriorityTaskWoken);
             }
         }
+
+        // Clear the interrupts after everything was read
+        uart->dev->int_clr.rxfifo_full = 1;
+        uart->dev->int_clr.frm_err = 1;
+        uart->dev->int_clr.rxfifo_tout = 1;
```

Comment:
there is a reason why interrupts are cleared BEFORE reading the bytes. Please revert.

Reply:
What is the reason? With all the respect that I have for you, trust me that is a lot since this Arduino framework is excellent, there are a lot of things like this that are not documented properly. 
Please comment the code properly to explain why the interrupts need to be cleared before reading the bytes

Reply:
there are cases where bytes might come in between the time you check and handle the bytes and the time you clear the interrupt. In that case you will not get an ISR for those bytes. We had that happen and scratched heads for quite some time. There was another case that I do not recall at this time as well. Generally your code had nothing to do with that, so I do not know why you would go and change the lines anyway :) I do accept the criticism that comment should be added for such cases.